src/test/java/hudson/plugins/git/GitSCMTest.java

```diff
@@ -290,6 +290,137 @@ public void testSpecificRefspecs() throws Exception {
         build(projectWithFoo, Result.SUCCESS, commitFile1);
     }
 
+    /**
+     * This test confirms the behaviour of avoiding the second fetch in GitSCM checkout()
+     **/
+    @Test
+    @Issue("JENKINS-56404")
+    public void testAvoidRedundantFetch() throws Exception {
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        repos.add(new UserRemoteConfig(testRepo.gitDir.getAbsolutePath(), "origin", "+refs/heads/*:refs/remotes/*", null));
+
+        /* Without honor refspec on initial clone */
+        FreeStyleProject projectWithMaster = setupProject(repos, Collections.singletonList(new BranchSpec("master")), null, false, null);
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master");
+
+        FreeStyleBuild build = build(projectWithMaster, Result.SUCCESS);
+
+        assertRedundantFetchIsTrue(build, "+refs/heads/*:refs/remotes/origin/*");
+    }
+
+    /**
+     * After avoiding the second fetch call in retrieveChanges(), this test verifies there is no data loss by fetching a repository
+     * (git init + git fetch) with a narrow refspec but without CloneOption of honorRefspec = true on initial clone
+     * First fetch -> wide refspec
+     * Second fetch -> narrow refspec (avoided)
+     **/
+    @Test
+    @Issue("JENKINS-56404")
+    public void testAvoidRedundantFetchWithoutHonorRefSpec() throws Exception {
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        repos.add(new UserRemoteConfig(testRepo.gitDir.getAbsolutePath(), "origin", "+refs/heads/foo:refs/remotes/foo", null));
+
+        /* Without honor refspec on initial clone */
+        FreeStyleProject projectWithMaster = setupProject(repos, Collections.singletonList(new BranchSpec("master")), null, false, null);
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master");
+        // Add another branch 'foo'
+        git.branch("foo");
+        git.checkout().branch("foo");
+        commit(commitFile1, johnDoe, "Commit in foo");
+
+        // Build will be success because the initial clone disregards refspec and fetches all branches
+        FreeStyleBuild build = build(projectWithMaster, Result.SUCCESS);
+        FilePath childFile = returnFile(build);
+
+        if (childFile != null) {
+            // assert that no data is lost by avoidance of second fetch
+            assertThat("master branch was not fetched", childFile.readToString(), containsString("master"));
+            assertThat("foo branch was not fetched", childFile.readToString(), containsString("foo"));
+        }
```

Comment:
Why should it be null? I might be misunderstanding the test, but I'd expect
```suggestion
        assertNotNull(childFile);
        // assert that no data is lost by avoidance of second fetch
        assertThat("master branch was not fetched", childFile.readToString(), containsString("master"));
        assertThat("foo branch was not fetched", childFile.readToString(), containsString("foo"));
```
This way we can be 100% sure that `childField` exists and there is no data loss. With the null check, it's possible that the block code under `childFile` is never executed, so you cannot be totally sure of what you're testing.

(it might be missing the assertNotNull import, beware)

Reply:
I just checked the method returning childFile, it can never return a null value. It wanted to check for an empty file, I'll create the relevant check, thanks!